### ISSUE TITLE
How to run a service

### ISSUE THREAD

**Issue by aiqency** [hashicorp/nomad-skeleton-driver-plugin#12]
What would be the easiest way to turn this driver plugin into a service?

I did a simple test by changing the job configuration from `batch` to `service` and putting the process to sleep for 10 seconds in order to simulate a running service.

```hcl
job "example" {
  datacenters = ["dc1"]
  type        = "service"  # change from batch to service
  ...
```

```golang
func (d *HelloDriverPlugin) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drivers.DriverNetwork, error) {
...
execCmd := &executor.ExecCommand{
	Args:       []string{"-c", `sleep 10`},  // fake 10 seconds running service
  ...
```

The task runs and is rescheduled after completion, with exit code 0, as if the nomad client did not report the correct status.

**I have several questions:**

Why the nomad plugin wait for the executor to terminate in two places:

```golang
func (d *HelloDriverPlugin) handleWait(ctx context.Context, ...) {
...
    h.exec.Wait(ctx)
...
}

func (h *taskHandle) run() {
...
    h.exec.Wait(context.Background())
...
}
```

Who/When `InspectTask` get called? As this seems to be the only method that cares about the `TaskStatus`.
```golang
func (h *taskHandle) TaskStatus() *drivers.TaskStatus 
```

Simply put, I would like to get an overview of how `services` should be performed and completed.

Thanks in advance.

**Comment by lgfa29** [comment]
Hi @aiqency 👋 

The job type is independent from the task driver being used. This skeleton plugin is just an example, so it's hard coded to just print a string, which is not useful for a `service` job.

If you are planning on writing your own plugin, you need to make sure it takes proper user input instead of hard coded commands, and start whatever sub-process is more adequate for the runtime you are targeting.

> Why the nomad plugin wait for the executor to terminate in two places:

[`exec.Wait`](https://pkg.go.dev/github.com/hashicorp/nomad/drivers/shared/executor#UniversalExecutor.Wait) just blocks until the process exits. In this sample plugin we call it in `handleWait` to notify Nomad when the process exits and in `run` to update the plugin's internal state.

> Who/When InspectTask get called? As this seems to be the only method that cares about the TaskStatus.

`InspectTask` is part of the [task driver plugin interface](https://www.nomadproject.io/docs/internals/plugins/task-drivers#task-driver-plugin-api), so it's called by Nomad.

> Simply put, I would like to get an overview of how services should be performed and completed.

This is outside the scope of the plugin. The difference between a `batch`  or a `service` job is how Nomad will handle the task lifecycle, so the plugin doesn't need to care (or even know) about this.

Plugins are responsible for starting, stopping, inspecting etc. a specific task. Nomad will call the plugin when necessary, depending on the job type and status.

 ----

Could provide some details on what you are trying to do? Maybe one of the existing task drivers would be fit your need instead of you having to write a custom one 🙂

**Comment by aiqency** [comment]
Hey @lgfa29, thank you very much for your answer.

> Could provide some details on what you are trying to do? Maybe one of the existing task drivers would be fit your need instead of you having to write a custom one

We already have on inter an application which can spawn vm's / containers from cli / REST api. We want nomad to be in charge of creating / managing these resources.

`Start/Stop` and `Exec` was not an issue thanks to this repo 👍, however, I realized that killing the process outside of the nomad context left the task status unchanged (running).

The most easy way to reproduce my issue is to call `sleep <whatever>` since my implementation so far is nothing more than a blocking process waiting for termination. 

> exec.Wait just blocks until the process exits. In this sample plugin we call it in handleWait to notify Nomad when the process exits and in run to update the plugin's internal state.

Can you confirm that only `InspectTask` consume the changes made by the `run()` method?

> The difference between a batch or a service job is how Nomad will handle the task lifecycle, so the plugin doesn't need to care (or even know) about this.

Then I'm even more confuse, why a simple `sleep 10` completes successfully in `batch` while failing when run as a `service`?

**Comment by lgfa29** [comment]
> however, I realized that killing the process outside of the nomad context left the task status unchanged (running).

The task driver plugin is responsible to monitor this, since Nomad doesn't know how to monitor external processes. This is done in the `WaitTask`function. In this skeleton example, the underlying library we're using (`github.com/hashicorp/nomad/drivers/shared/executor`) provides a blocking channel that we can monitor to detect when the process exits.

In your case, you will need to find a way to do this. For example, you could store the process ID in your driver state and then monitor it. If it stops running your plugin can notify Nomad.

> Can you confirm that only InspectTask consume the changes made by the run() method?

Hum....I'm not sure what you mean by this. `run()` is called by `StartTask` and updates the task state. This state is used by the plugin in several places (essentially anywhere `handle.*` is called).

> Then I'm even more confuse, why a simple sleep 10 completes successfully in batch while failing when run as a service?

Because a service job is expected to run continuously and not exit. If it exits, Nomad will consider it failed and restart. A batch job, on the other hand, is expected to run to completion, so if if exits, that's expected. Checkout our [schedulers docs](https://www.nomadproject.io/docs/schedulers) for more info.

**Comment by aiqency** [comment]
Thank very much, I think you've given me enough clues to continue.

> For example, you could store the process ID in your driver state and then monitor it. If it stops running your plugin can notify Nomad.

I will probably take an approach similar to this yes.

Much appreciated 👍.

**Comment by lgfa29** [comment]
Nice, feel free to reach out if you have any more questions.

You can check the Podman task driver for an example. [`WaitTask` starts a goroutine](https://github.com/hashicorp/nomad-driver-podman/blob/v0.3.0/driver.go#L737) to monitor if the containers are still running, and that goroutine [exits](https://github.com/hashicorp/nomad-driver-podman/blob/v0.3.0/handle.go#L83-L85) when the container stops running for any reason (failed, crashed, killed outside Nomad etc.) There's also [another goroutine](https://github.com/hashicorp/nomad-driver-podman/blob/v0.3.0/handle.go#L193) that monitors and updates the internal state of the plugin.

**Comment by aiqency** [comment]
> Nice, feel free to reach out if you have any more questions.

Hi @lgfa29, I have a [pending question](https://discuss.hashicorp.com/t/nomad-check-on-custom-plugin/29107) on discuss.hashicorp, hope you can take a look.
